cloud/services/resourceskus/cache.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourceskus
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-01/compute"
+	"github.com/pkg/errors"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+)
+
+// Cache loads resource SKUs at the beginning of reconcile to expose
+// features available on compute resources. It exposes convenience
+// functionality for trawling Azure SKU capabilities. It may be adapted
+// to periodically refresh data in the background.
+type Cache struct {
+	client Client
+
+	// data is the cached sku information from Azure.
+	// synchronization required if data is cached across reconcile calls, (i.e., refreshed in background as Runnable via mgr.Add(...))
+	data []compute.ResourceSku
+}
+
+// NewCacheFunc allows for mocking out the underlying client
+type NewCacheFunc func(context.Context, azure.Authorizer) (*Cache, error)
+
+// NewCache instantiates a cache and initializes its contents.
+func NewCache(ctx context.Context, auth azure.Authorizer) (*Cache, error) {
+	cache := &Cache{
+		client: NewClient(auth),
+	}
+
+	if err := cache.refresh(ctx); err != nil {
+		return nil, err
+	}
+
+	return cache, nil
+}
+
+// NewStaticCacheFn returns a function that initializes a cache with data and no ability to refresh. Used for testing.
+func NewStaticCacheFn(data []compute.ResourceSku) NewCacheFunc {
+	return func(context.Context, azure.Authorizer) (*Cache, error) {
+		return NewStaticCache(data), nil
+	}
+}
+
+// NewStaticCache initializes a cache with data and no ability to refresh. Used for testing.
+func NewStaticCache(data []compute.ResourceSku) *Cache {
+	return &Cache{
+		data: data,
+	}
+}
+
+func (c *Cache) refresh(ctx context.Context) error {
+	data, err := c.client.List(ctx, "")
```

Comment:
Hmm... Would it be better to scope the query by location?

Reply:
exactly my other comment heh

I suspect this will greatly reduce the size, I forgot we stopped doing this